WebKit should not depend on CA's implicit auto-upgrade to extendedSRGB

hortont424 opened the pull request:
#### 831c68acbcc6c5c7992a250059b330f261bcf2a0
<pre>
WebKit should not depend on CA&apos;s implicit auto-upgrade to extendedSRGB
<a href="https://bugs.webkit.org/show_bug.cgi?id=247218">https://bugs.webkit.org/show_bug.cgi?id=247218</a>
rdar://101695786

Reviewed by Said Abou-Hallawa and Sam Weinig.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::colorSpace const):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
We currently depend on a special case in CA to upgrade RemoteLayerBackingStore&apos;s
backing store to extended SRGB in the case where we are using half-float pixel
formats. We ask for standard SRGB, but they notice the floating point format
and upgrade us.

Instead of doing this, we should explicitly say that we want extended SRGB
when using such backing store.

Canonical link: <a href="https://commits.webkit.org/256361@main">https://commits.webkit.org/256361@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d094b9236ab19583bdab325f44ed6a4dd253209c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95527 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4792 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105107 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4822 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33525 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87897 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100958 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101190 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3516 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82138 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30600 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87322 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73435 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener, /TestWTF:WTF_ParkingLot.UnparkOneOneFast (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39279 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36980 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20178 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4395 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40976 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42964 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39429 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -187,6 +187,15 @@
+DestinationColorSpace RemoteLayerBackingStore::colorSpace() const
+{
+#if HAVE(IOSURFACE_RGB10)
+    if (m_parameters.type == Type::IOSurface && m_parameters.deepColor)
```
Can we have the `#if HAVE(..)` and this condition in a helper function since it is repeated in `RemoteLayerBackingStore::pixelFormat()` also? e.g. `isDeepColorBackingStore()`, `isHalfFloatPixelFormatSupported()`,` isRGB10PixelFormatSupported()` or something else.

hortont424 replied:
An alternative probably-even-better approach is to adopt `PixelBufferFormat` (and maybe even thread it through ImageBuffer::create too)

hortont424 replied:
(and then we only have one function again)

weinig commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
```diff
@@ -187,6 +187,15 @@
+{
+#if HAVE(IOSURFACE_RGB10)
+    if (m_parameters.type == Type::IOSurface && m_parameters.deepColor)
+        return DestinationColorSpace { extendedSRGBColorSpaceRef() };
```
If you are feeling motivated, probably makes sense to add a static on DestinationColorSpace for this, as its used in a few places now.

hortont424 committed:
WebKit should not depend on CA's implicit auto-upgrade to extendedSRGB
https://bugs.webkit.org/show_bug.cgi?id=247218
rdar://101695786

Reviewed by Said Abou-Hallawa and Sam Weinig.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::colorSpace const):
(WebKit::RemoteLayerBackingStore::ensureFrontBuffer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
We currently depend on a special case in CA to upgrade RemoteLayerBackingStore's
backing store to extended SRGB in the case where we are using half-float pixel
formats. We ask for standard SRGB, but they notice the floating point format
and upgrade us.

Instead of doing this, we should explicitly say that we want extended SRGB
when using such backing store.

Canonical link: https://commits.webkit.org/256361@main
